src/h3lib/lib/h3Index.c

```diff
@@ -78,49 +78,108 @@ void H3_EXPORT(h3ToString)(H3Index h, char *str, size_t sz) {
-            foundFirstNonZeroDigit = true;
-            if (_isBaseCellPentagon(baseCell) && digit == K_AXES_DIGIT) {
+    // Implementation strategy:
+    //
+    // Walk from high to low bits, checking validity of
+    // groups of bits as we go. After each check, shift bits off
+    // to the left, so that the next relevant group is at the
+    // highest bit location in the H3 Index, which we can
+    // easily read off with the `GT` macro. This strategy helps
+    // us avoid re-computing shifts and masks for each group.
+    //
+    // |   Region   | # bits |
+    // |------------|--------|
+    // | High       |      1 |
+    // | Mode       |      4 |
+    // | Reserved   |      3 |
+    // | Resolution |      4 |
+    // | Base Cell  |      7 |
+    // | Digit 1    |      3 |
+    // | Digit 2    |      3 |
+    // | ...        |    ... |
+    // | Digit 15   |      3 |
+    //
+    // Additionally, we try to group operations and void loops when possible.
+
+    // The 1 high bit should be 0b0
+    // The 4 mode bits should be 0b0001 (H3_CELL_MODE)
+    // The 3 reserved bits should be 0b000
+    // In total, the top 8 bits should be 0b00001000
+    if (GT(h, 8) != 0b00001000) return 0;
+    h <<= 8;
+
+    // Number of bits in each of the resolution, base cell, and digit regions.
+    const int nBitsRes = 4;
+    const int nBitsBaseCell = 7;
+    const int nBitsDigit = H3_PER_DIGIT_OFFSET;
+
+    // No need to check resolution; any 4-bit number is a valid resolution.
+    const int res = GT(h, nBitsRes);
+    h <<= nBitsRes;
+
+    // Check that base cell number is valid.
+    const int bc = GT(h, nBitsBaseCell);
+    if (bc >= NUM_BASE_CELLS) return 0;
+    h <<= nBitsBaseCell;
+
+    // Now check that each resolution digit is valid.
+    // Let `r` denote the resolution we're currently checking.
+    int r = 1;
+
+    // Pentagon cells start with a sequence of 0's (CENTER_DIGIT's).
+    // The first nonzero digit can't be a 1 (i.e., "deleted subsequence",
+    // PENTAGON_SKIPPED_DIGIT, or K_AXES_DIGIT).
+    // Test for pentagon base cell first to avoid this loop if possible.
+    if (isBaseCellPentagonArr[bc]) {
+        for (; r <= res; r++) {
+            int d = GT(h, nBitsDigit);
+            if (d == 0) {
+                h <<= nBitsDigit;
+            } else if (d == 1) {
                 return 0;
+            } else {
+                break;
+                // But don't increment `r`, since we still need to
+                // check that it isn't a 7.
             }
         }
-
-        if (digit < CENTER_DIGIT || digit >= NUM_DIGITS) return 0;
     }
 
-    for (int r = res + 1; r <= MAX_H3_RES; r++) {
-        Direction digit = H3_GET_INDEX_DIGIT(h, r);
-        if (digit != INVALID_DIGIT) return 0;
+    // After (possibly) taking care of pentagon logic, check that
+    // the remaining digits up to `res` are not 7 (INVALID_DIGIT).
+    // Don't see a way to avoid this loop :(
```

Comment:
You could unroll it into groups of 4 or 8, possibly with a macro to simplify the code. This would only work for hex base cells though, or for pentagon base cells if you re-check all the digits you just checked. Might be worth trying.

Reply:
Not sure if this PR is still alive, but I've just stumbled on it.

Last week I've implemented my own quick `isValidCell` (a simpler one though, because for my use case I don't need to have any specific check for pentagons) and I think this loop can be avoided.

The search for the pattern `0b111` may not be easy to implement quickly, but on the other hand we can apply a bitwise NOT and look for `0b000` which is easier 🙂.
Indeed, looking for a null triplet is similar to looking for a nul-byte, something we do a lot in C (think `strlen`).

By digging into the annals of the Old Gods, a.k.a. `comp.lang.c`, we can extract this little gem: [Alan Mycroft's null-byte detection algorithm](https://groups.google.com/forum/#!original/comp.lang.c/2HtQXvg7iKc/xOJeipH6KLMJ), posted in 1987.

Tweaking the bitmasks to works on 3-bit instead of 8-bit does the trick and we end up with a code that would rougly look like

```c
#include<stdint.h>
#include<stdio.h>
#include<assert.h>

#define MAX_RESOLUTION 15
#define CELL_BITSIZE   3


int has_invalid_digit(uint64_t index) {
    static const uint64_t LO_MAGIC = 0x49249249249ULL;  // ...001 001 001...
    static const uint64_t HI_MAGIC = 0x124924924924ULL; // ...100 100 100...

    const uint64_t resolution = (index >> 52 & 0xF);
    const uint64_t unused_count = MAX_RESOLUTION - resolution;
    const uint64_t unused_bitsize = unused_count * CELL_BITSIZE;
    const uint64_t digits_mask = (1ULL << (resolution * CELL_BITSIZE)) - 1;
    const uint64_t digits = (index >> unused_bitsize) & digits_mask;

    return ((~digits - LO_MAGIC) & (digits & HI_MAGIC)) != 0ULL;
}

int main(void) {
    // Valid H3 index
    // 0-0001-000-1100-0010101-110-101-110-001-100-000-101-001-100-110-110-101-111-111-111
    assert(!has_invalid_digit(0x08c2bae305336bffULL));
    // First digit invalid.
    // 0-0001-000-1100-0010101-111-101-110-001-100-000-101-001-100-110-110-101-111-111-111
    assert(has_invalid_digit(0x08c2bee305336bffULL));
    // Digit in the middle invalid.
    // 0-0001-000-1100-0010101-110-101-110-001-100-111-101-001-100-110-110-101-111-111-111
    assert(has_invalid_digit(0x08c2bae33d336bffULL));
    // Last digit invalid.
    // 0-0001-000-1100-0010101-110-101-110-001-100-000-101-001-100-110-110-111-111-111-111
    assert(has_invalid_digit(0x08c2bae305336fffULL));

    return 0;
}
```

So, in the end we can replace the loop by `((~digits - LO_MAGIC) & (digits & HI_MAGIC)) != 0ULL`.

For sure, this would deserve some nice comments, because it's not obvious on a first read (I went heavy on the comments for my implementation xD).

If something is not clear I can give a more detailed explanation (but that basically boils down on how Alan Mycroft's null-byte detection works).

Reply:
Wow, this is awesome! What a great find!

I spent a little time reworking this PR to add your logic. It definitely helps on the benchmarks.

It looks like you also came up with some improvements for the pentagon branch here: https://github.com/nmandery/h3ron/issues/34

I'll have to parse that as well. Also happy if you want to make a pull against this PR!